File: backend/src/application/agent/LlmOptimizationProvider.ts
import { Environment } from "../../config/Environment.js";
import { LlmProviderConfig } from "../../domain/entities/LlmProviderConfig.js";
import { OptimizationDraft } from "./OptimizationDraft.js";
import {
  OptimizationProvider,
  OptimizationProviderRequest,
  OptimizationProviderResult
} from "./OptimizationProvider.js";

export class LlmOptimizationProvider implements OptimizationProvider {
  readonly providerName: string;

  constructor(
    private readonly environment: Environment,
    private readonly config: LlmProviderConfig
  ) {
    this.providerName = config.key;
  }

  async generate(request: OptimizationProviderRequest): Promise<OptimizationProviderResult | null> {
    const apiUrl = this.config.baseUrl || this.environment.llmApiUrl;

    if (!apiUrl) {
      return null;
    }

    const prompt = request.prompt;
    const response = await fetch(apiUrl, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        model: this.config.model,
        messages: [
          {
            role: "system",
            content: "You are a precise ecommerce merchandising assistant for Shopify product listings."
          },
          { role: "user", content: prompt }
        ],
        response_format: request.stepName === "final_synthesis" ? { type: "json_object" } : undefined
      })
    });

    if (!response.ok) {
      return null;
    }

    const body = (await response.json()) as unknown;
    const text = this.toText(body);

    if (!text) {
      return null;
    }

    const usage = this.toUsage(body, prompt, text);
    return {
      response: text,
      draft: request.stepName === "final_synthesis" ? this.toDraft(text) : null,
      provider: this.providerName,
      model: this.config.model,
      inputTokens: usage.inputTokens,
      outputTokens: usage.outputTokens,
      estimatedCost: this.estimateCost(usage.inputTokens, usage.outputTokens)
    };
  }

  private headers(): Record<string, string> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    const configuredKey = this.config.apiKeyEnvVar ? this.environment.getOptional(this.config.apiKeyEnvVar) : "";
    const apiKey = configuredKey || this.environment.llmApiKey;

    if (apiKey) {
      headers.Authorization = `Bearer ${apiKey}`;
    }

    return headers;
  }

  private toText(body: unknown): string | null {
    if (!body || typeof body !== "object") {
      return null;
    }

    const payload = body as Record<string, unknown>;
    const choices = payload.choices;

    if (Array.isArray(choices)) {
      const first = choices[0] as Record<string, unknown> | undefined;
      const message = first?.message as Record<string, unknown> | undefined;
      const content = message?.content ?? first?.text;
      return typeof content === "string" ? content : null;
    }

    if (typeof payload.output_text === "string") {
      return payload.output_text;
    }

    if (typeof payload.content === "string") {
      return payload.content;
    }

    return null;
  }

  private toUsage(body: unknown, prompt: string, response: string): { inputTokens: number; outputTokens: number } {
    const payload = body && typeof body === "object" ? (body as Record<string, unknown>) : {};
    const usage = payload.usage && typeof payload.usage === "object"
      ? (payload.usage as Record<string, unknown>)
      : {};

    return {
      inputTokens: this.numberFrom(usage.prompt_tokens ?? usage.input_tokens) ?? this.estimateTokens(prompt),
      outputTokens: this.numberFrom(usage.completion_tokens ?? usage.output_tokens) ?? this.estimateTokens(response)
    };
  }

  private toDraft(text: string): OptimizationDraft | null {
    const candidate = this.parseJsonObject(text);

    if (!candidate || typeof candidate.improvedDescription !== "string") {
      return null;
    }

    return {
      improvedTitle: typeof candidate.improvedTitle === "string" ? candidate.improvedTitle : null,
      improvedDescription: candidate.improvedDescription,
      tags: Array.isArray(candidate.tags) ? candidate.tags.map(String) : [],
      positioningRecommendations:
        typeof candidate.positioningRecommendations === "string" ? candidate.positioningRecommendations : null,
      seoNotes: typeof candidate.seoNotes === "string" ? candidate.seoNotes : null,
      score: typeof candidate.score === "number" ? candidate.score : null,
      rationale: typeof candidate.rationale === "string" ? candidate.rationale : null
    };
  }

  private parseJsonObject(text: string): Record<string, unknown> | null {
    const trimmed = text.trim();
    const json = trimmed.startsWith("{")
      ? trimmed
      : trimmed.slice(trimmed.indexOf("{"), trimmed.lastIndexOf("}") + 1);

    if (!json) {
      return null;
    }

    try {
      const parsed = JSON.parse(json) as unknown;
      return parsed && typeof parsed === "object" && !Array.isArray(parsed)
        ? (parsed as Record<string, unknown>)
        : null;
    } catch {
      return null;
    }
  }

  private estimateTokens(value: string): number {
    return Math.max(1, Math.ceil(value.length / 4));
  }

  private estimateCost(inputTokens: number, outputTokens: number): number {
    const inputCost = (inputTokens / 1000) * this.config.inputCostPer1K;
    const outputCost = (outputTokens / 1000) * this.config.outputCostPer1K;
    return Number((inputCost + outputCost).toFixed(6));
  }

  private numberFrom(value: unknown): number | null {
    return typeof value === "number" && Number.isFinite(value) ? value : null;
  }
}

File: app/services/agent/llm_optimization_provider.rb
require "net/http"
require "json"

module Agent
  class LlmOptimizationProvider
    def initialize(config)
      @config = config
    end

    def generate(step_name:, prompt:, snapshot:, analysis:, prior_responses: {})
      api_url = @config.base_url.presence || ENV.fetch("LLM_API_URL", "")
      return nil if api_url.blank?

      body = build_body(step_name, prompt)
      response = post(api_url, body)
      return nil unless response

      text = extract_text(response)
      return nil if text.nil?

      usage = extract_usage(response, prompt, text)
      {
        response: text,
        draft: step_name == "final_synthesis" ? parse_draft(text) : nil,
        provider: @config.key,
        model: @config.model,
        input_tokens: usage[:input_tokens],
        output_tokens: usage[:output_tokens],
        estimated_cost: estimate_cost(usage[:input_tokens], usage[:output_tokens])
      }
    rescue => e
      Rails.logger.warn("[LlmOptimizationProvider] request failed: #{e.message}")
      nil
    end

    private

    def build_body(step_name, prompt)
      body = {
        model: @config.model,
        messages: [
          { role: "system", content: "You are a precise ecommerce merchandising assistant for Shopify product listings." },
          { role: "user", content: prompt }
        ]
      }
      body[:response_format] = { type: "json_object" } if step_name == "final_synthesis"
      body
    end

    def post(api_url, body)
      uri  = URI(api_url)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.open_timeout = 10
      http.read_timeout = 60

      req = Net::HTTP::Post.new(uri)
      req["Content-Type"] = "application/json"
      api_key = resolved_api_key
      req["Authorization"] = "Bearer #{api_key}" if api_key.present?
      req.body = body.to_json

      res = http.request(req)
      res.is_a?(Net::HTTPSuccess) ? JSON.parse(res.body) : nil
    rescue Net::OpenTimeout, Net::ReadTimeout, Errno::ECONNREFUSED => e
      Rails.logger.warn("[LlmOptimizationProvider] connection error: #{e.message}")
      nil
    end

    def extract_text(body)
      choices = body["choices"]
      if choices.is_a?(Array)
        first = choices.first || {}
        content = first.dig("message", "content") || first["text"]
        return content if content.is_a?(String)
      end

      return body["output_text"] if body["output_text"].is_a?(String)
      return body["content"]     if body["content"].is_a?(String)
      nil
    end

    def extract_usage(body, prompt, response_text)
      usage = body["usage"] || {}
      {
        input_tokens:  (usage["prompt_tokens"] || usage["input_tokens"])&.to_i     || estimate_tokens(prompt),
        output_tokens: (usage["completion_tokens"] || usage["output_tokens"])&.to_i || estimate_tokens(response_text)
      }
    end

    def parse_draft(text)
      trimmed = text.strip
      json_str = if trimmed.start_with?("{")
        trimmed
      else
        start_i = trimmed.index("{")
        end_i   = trimmed.rindex("}")
        start_i && end_i ? trimmed[start_i..end_i] : nil
      end
      return nil unless json_str

      parsed = JSON.parse(json_str)
      return nil unless parsed.is_a?(Hash) && parsed["improvedDescription"].is_a?(String)

      {
        improved_title:                parsed["improvedTitle"]&.is_a?(String) ? parsed["improvedTitle"] : nil,
        improved_description:          parsed["improvedDescription"],
        tags:                          Array(parsed["tags"]).map(&:to_s),
        positioning_recommendations:   parsed["positioningRecommendations"]&.is_a?(String) ? parsed["positioningRecommendations"] : nil,
        seo_notes:                     parsed["seoNotes"]&.is_a?(String) ? parsed["seoNotes"] : nil,
        score:                         parsed["score"]&.is_a?(Numeric) ? parsed["score"] : nil,
        rationale:                     parsed["rationale"]&.is_a?(String) ? parsed["rationale"] : nil
      }
    rescue JSON::ParserError
      nil
    end

    def estimate_cost(input_tokens, output_tokens)
      input_cost  = (input_tokens  / 1000.0) * @config.input_cost_per1k.to_f
      output_cost = (output_tokens / 1000.0) * @config.output_cost_per1k.to_f
      (input_cost + output_cost).round(6)
    end

    def estimate_tokens(text)
      [(text.length / 4.0).ceil, 1].max
    end

    def resolved_api_key
      env_var = @config.api_key_env_var.presence
      (env_var ? ENV[env_var].presence : nil) || ENV.fetch("LLM_API_KEY", "")
    end
  end
end
